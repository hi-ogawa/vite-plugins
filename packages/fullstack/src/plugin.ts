import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import MagicString from "magic-string";
import { toNodeHandler } from "srvx/node";
import {
  DevEnvironment,
  type Plugin,
  type ResolvedConfig,
  type Rollup,
  type ViteDevServer,
  isCSSRequest,
  isRunnableDevEnvironment,
  normalizePath,
} from "vite";
import type { ImportAssetsOptions, ImportAssetsResult } from "../types/shared";
import {
  type AssetsVirtual,
  parseAssetsVirtual,
  toAssetsVirtual,
} from "./plugins/shared";
import {
  createVirtualPlugin,
  getEntrySource,
  hashString,
  normalizeRelativePath,
} from "./plugins/utils";
import {
  evalValue,
  normalizeViteImportAnalysisUrl,
} from "./plugins/vite-utils";

type FullstackPluginOptions = {
  /**
   * @default true
   */
  serverHandler?: boolean;
  /**
   * @default ["ssr"]
   */
  serverEnvironments?: string[];
};

type ImportAssetsMeta = {
  id: string;
  key: string;
  importerEnvironment: string;
  virtual: AssetsVirtual;
};

export default function vitePluginFullstack(
  pluginOpts?: FullstackPluginOptions,
): Plugin[] {
  return [...serverHandlerPlugin(pluginOpts), ...assetsPlugin(pluginOpts)];
}

export function serverHandlerPlugin(
  pluginOpts?: FullstackPluginOptions,
): Plugin[] {
  return [
    {
      name: "fullstack:server-handler",
      apply: () => pluginOpts?.serverHandler !== false,
      config(userConfig, _env) {
        return {
          appType: userConfig.appType ?? "custom",
        };
      },
      configureServer(server) {
        const name = (pluginOpts?.serverEnvironments ?? ["ssr"])[0]!;
        const environment = server.environments[name]!;
        assert(isRunnableDevEnvironment(environment));
        const runner = environment.runner;
        return () => {
          server.middlewares.use(async (req, res, next) => {
            try {
              const source = getEntrySource(environment.config);
              const mod = await runner.import(source);
              req.url = req.originalUrl ?? req.url;
              await toNodeHandler(mod.default.fetch)(req, res);
            } catch (e) {
              next(e);
            }
          });
        };
      },
    },
  ];
}

export function assetsPlugin(pluginOpts?: FullstackPluginOptions): Plugin[] {
  let server: ViteDevServer;
  let resolvedConfig: ResolvedConfig;
  const importAssetsMetaMap: {
    [environment: string]: { [id: string]: ImportAssetsMeta };
  } = {};
  const bundleMap: { [environment: string]: Rollup.OutputBundle } = {};

  return [
    {
      name: "fullstack:assets",
      // TODO: support non shared build?
      sharedDuringBuild: true,
      configureServer(server_) {
        server = server_;
      },
      configResolved(config) {
        resolvedConfig = config;
      },
      configEnvironment(name) {
        const serverEnvironments = pluginOpts?.serverEnvironments ?? ["ssr"];
        if (serverEnvironments.includes(name)) {
          return {
            build: {
              emitAssets: true,
            },
          };
        }
      },
      /**
       * [Transform input]
       *   const assets = import.meta.vite.assets(...)
       *
       * [Transform output]
       *   import __assets_xxx from "virtual:fullstack/assets?..."
       *   const assets = __assets_xxx
       */
      transform: {
        async handler(code, id, _options) {
          if (!code.includes("import.meta.vite.assets")) return;

          // TODO: strip comments
          const output = new MagicString(code);

          const emptyResult: ImportAssetsResult = {
            js: [],
            css: [],
          };

          const newImports = new Set<string>();

          for (const match of code.matchAll(
            /import\.meta\.vite\.assets\(([\s\S]*?)\)/dg,
          )) {
            const [start, end] = match.indices![0]!;

            // No-op on client since vite build handles preload/css for dynamic import on client.
            // https://vite.dev/guide/features.html#async-chunk-loading-optimization
            if (this.environment.name === "client") {
              const replacement = `(${JSON.stringify(emptyResult)})`;
              output.update(start, end, replacement);
              continue;
            }

            const argCode = match[1]!.trim();
            const options: Required<ImportAssetsOptions> = {
              import: id,
              environment: this.environment.name,
              asEntry: false,
            };
            if (argCode) {
              const argValue = evalValue<ImportAssetsOptions>(argCode);
              Object.assign(options, argValue);
            }

            const importSource = toAssetsVirtual({
              import: options.import,
              importer: id,
              environment: options.environment,
              entry: options.asEntry ? "1" : "",
            });
            const hash = hashString(importSource);
            const importedName = `__assets_${hash}`;
            newImports.add(
              `;import ${importedName} from ${JSON.stringify(importSource)};\n`,
            );
            output.update(start, end, `(${importedName})`);
          }

          if (output.hasChanged()) {
            // add virtual imports at the end so that other imports are already processed
            // and css already exists in server module graph.
            // TODO: forgot to do this on `@vitejs/plugin-rsc`
            // TODO: nop, we probably want to eagerly transform anyways.
            for (const newImport of newImports) {
              output.append(newImport);
            }
            return {
              code: output.toString(),
              map: output.generateMap({ hires: "boundary" }),
            };
          }
        },
      },
      resolveId: {
        handler(source) {
          if (source.startsWith("virtual:fullstack/assets?")) {
            return "\0" + source;
          }
          if (source === "virtual:fullstack/assets-manifest") {
            assert.notEqual(this.environment.name, "client");
            assert.equal(this.environment.mode, "build");
            return { id: source, external: true };
          }
        },
      },
      load: {
        async handler(id) {
          const parsed = parseAssetsVirtual(id);
          if (!parsed) return;
          assert.notEqual(this.environment.name, "client");

          // TODO: we probably shouldn't resolve client file on other environment.
          // we could avoid this by another virtual but it's possible only for dev.
          const resolved = await this.resolve(parsed.import, parsed.importer);
          assert(resolved, `Failed to resolve: ${parsed.import}`);

          if (this.environment.mode === "dev") {
            const result: ImportAssetsResult = {
              entry: undefined, // defined only on client
              js: [], // always empty
              css: [], // defined only on server
            };
            const environment = server.environments[parsed.environment];
            assert(environment, `Unknown environment: ${parsed.environment}`);
            if (parsed.environment === "client") {
              result.entry = normalizeViteImportAnalysisUrl(
                environment,
                resolved.id,
              );
            }
            if (environment.name !== "client") {
              const collected = await collectCss(environment, resolved.id);
              for (const file of [resolved.id, ...collected.visitedFiles]) {
                if (fs.existsSync(file)) {
                  this.addWatchFile(file);
                }
              }
              result.css = collected.hrefs.map((href, i) => ({
                href,
                "data-vite-dev-id": collected.ids[i],
              }));
            }
            return `export default ${JSON.stringify(result)}`;
          } else {
            const meta: ImportAssetsMeta = {
              id: resolved.id,
              // normalize to have machine-independent build output
              key: path.relative(resolvedConfig.root, resolved.id),
              importerEnvironment: this.environment.name,
              virtual: parsed,
            };
            // TODO: don't overwrite entry=true to entry=false
            (importAssetsMetaMap[parsed.environment] ??= {})[meta.id] = meta;
            return `\
              import __assets_manifest from "virtual:fullstack/assets-manifest";
              export default __assets_manifest[${JSON.stringify(parsed.environment)}][${JSON.stringify(meta.key)}];
            `;
          }
        },
      },
      // non-client builds can load assets manifest as external
      renderChunk(code, chunk) {
        if (code.includes("virtual:fullstack/assets-manifest")) {
          const replacement = normalizeRelativePath(
            path.relative(
              path.join(chunk.fileName, ".."),
              BUILD_ASSETS_MANIFEST_NAME,
            ),
          );
          code = code.replaceAll(
            "virtual:fullstack/assets-manifest",
            () => replacement,
          );
          return { code };
        }
        return;
      },
      writeBundle(_options, bundle) {
        bundleMap[this.environment.name] = bundle;
      },
      buildStart() {
        // dynamically add client entry during build
        if (
          this.environment.mode == "build" &&
          this.environment.name === "client"
        ) {
          const metas = importAssetsMetaMap["client"];
          if (metas) {
            for (const meta of Object.values(importAssetsMetaMap["client"]!)) {
              if (meta.virtual.entry) {
                this.emitFile({
                  type: "chunk",
                  id: meta.id,
                });
              }
            }
          }
        }
      },
      buildApp: {
        order: "post",
        async handler(builder) {
          // build manifest of imported assets
          const manifest: BuildAssetsManifest = {};
          for (const [environmentName, metas] of Object.entries(
            importAssetsMetaMap,
          )) {
            const bundle = bundleMap[environmentName]!;
            const assetDepsMap = collectAssetDeps(bundle);
            for (const [id, meta] of Object.entries(metas)) {
              const found = assetDepsMap[id];
              if (!found) {
                return this.error(
                  `[vite-plugin-fullstack] failed to find built chunk for ${meta.id} imported by ${meta.importerEnvironment} environment`,
                );
              }
              const result: ImportAssetsResult = {
                js: [],
                css: [],
              };
              const { chunk, deps } = found;
              // TODO: base
              if (environmentName === "client") {
                result.entry = `/${chunk.fileName}`;
                result.js = deps.js.map((fileName) => ({
                  href: `/${fileName}`,
                }));
              }
              result.css = deps.css.map((fileName) => ({
                href: `/${fileName}`,
              }));
              (manifest[environmentName] ??= {})[meta.key] = result;
            }
          }

          // write manifest to importer environments
          const importerEnvironments = new Set(
            Object.values(importAssetsMetaMap)
              .flatMap((metas) => Object.values(metas))
              .flatMap((meta) => meta.importerEnvironment),
          );
          for (const environmentName of importerEnvironments) {
            const outDir =
              builder.environments[environmentName]!.config.build.outDir;
            fs.writeFileSync(
              path.join(outDir, BUILD_ASSETS_MANIFEST_NAME),
              `export default ${JSON.stringify(manifest, null, 2)};`,
            );

            // copy assets to client (mainly for server css)
            const clientOutDir =
              builder.environments["client"]!.config.build.outDir;
            for (const asset of Object.values(bundleMap[environmentName]!)) {
              if (asset.type === "asset") {
                const srcFile = path.join(outDir, asset.fileName);
                const destFile = path.join(clientOutDir, asset.fileName);
                fs.mkdirSync(path.dirname(destFile), { recursive: true });
                fs.copyFileSync(srcFile, destFile);
              }
            }
          }
        },
      },
    },
    // ensure at least one client build input to prevent Vite
    // from looking for index.html and breaking build
    {
      ...createVirtualPlugin("fullstack/client-fallback", () => "export {}"),
      configEnvironment: {
        order: "post",
        handler(name, config, _env) {
          if (name === "client") {
            if (!config.build?.rollupOptions?.input) {
              return {
                build: {
                  rollupOptions: {
                    input: {
                      __fallback: "virtual:fullstack/client-fallback",
                    },
                  },
                },
              };
            }
          }
        },
      },
      generateBundle(_optoins, bundle) {
        if (this.environment.name !== "client") return;
        for (const [k, v] of Object.entries(bundle)) {
          if (v.type === "chunk" && v.name === "__fallback") {
            delete bundle[k];
          }
        }
      },
    },
    patchViteClientPlugin(),
    patchVueScopeCssHmr(),
  ];
}

const BUILD_ASSETS_MANIFEST_NAME = "__fullstack_assets_manifest.js";

async function collectCss(environment: DevEnvironment, entryId: string) {
  const visited = new Set<string>();
  const cssIds = new Set<string>();
  const visitedFiles = new Set<string>();

  async function recurse(id: string) {
    if (visited.has(id) || parseAssetsVirtual(id)) {
      return;
    }
    visited.add(id);
    const mod = environment.moduleGraph.getModuleById(id);
    if (!mod) return;
    // TODO: this is not ideal.
    // We eagerly transforming module is required
    // to allow patterns like packages/fullstack/examples/react-router/src/routes.ts
    if (!mod.transformResult) {
      try {
        await environment.transformRequest(id);
      } catch (e) {
        console.error(`[collectCss] Failed to transform '${id}'`, e);
      }
    }
    if (mod?.file) {
      visitedFiles.add(mod.file);
    }
    for (const next of mod?.importedModules ?? []) {
      if (next.id) {
        if (isCSSRequest(next.id)) {
          if (hasSpecialCssQuery(next.id)) {
            continue;
          }
          cssIds.add(next.id);
        } else {
          recurse(next.id);
        }
      }
    }
  }

  await recurse(entryId);

  // this doesn't include ?t= query so that RSC <link /> won't keep adding styles.
  const hrefs = [...cssIds].map((id) =>
    normalizeViteImportAnalysisUrl(environment, id),
  );
  return { ids: [...cssIds], hrefs, visitedFiles: [...visitedFiles] };
}

function hasSpecialCssQuery(id: string): boolean {
  return /[?&](url|inline|raw)(\b|=|&|$)/.test(id);
}

type BuildAssetsManifest = {
  [environment: string]: {
    [import_: string]: ImportAssetsResult;
  };
};

type AssetDeps = {
  js: string[];
  css: string[];
};

type AssetDepsMap = {
  [id: string]: { chunk: Rollup.OutputChunk; deps: AssetDeps };
};

function collectAssetDeps(bundle: Rollup.OutputBundle) {
  const chunkToDeps = new Map<Rollup.OutputChunk, AssetDeps>();
  for (const chunk of Object.values(bundle)) {
    if (chunk.type === "chunk") {
      chunkToDeps.set(chunk, collectAssetDepsInner(chunk.fileName, bundle));
    }
  }
  const idToDeps: AssetDepsMap = {};
  for (const [chunk, deps] of chunkToDeps.entries()) {
    for (const id of chunk.moduleIds) {
      idToDeps[id] = { chunk, deps };
    }
  }
  return idToDeps;
}

function collectAssetDepsInner(
  fileName: string,
  bundle: Rollup.OutputBundle,
): AssetDeps {
  const visited = new Set<string>();
  const css: string[] = [];

  function recurse(k: string) {
    if (visited.has(k)) return;
    visited.add(k);
    const v = bundle[k];
    assert(v, `Not found '${k}' in the bundle`);
    if (v.type === "chunk") {
      css.push(...(v.viteMetadata?.importedCss ?? []));
      for (const k2 of v.imports) {
        // server external imports is not in bundle
        if (k2 in bundle) {
          recurse(k2);
        }
      }
    }
  }

  recurse(fileName);
  return {
    js: [...visited],
    css: [...new Set(css)],
  };
}

// TODO: patch @vite/client for https://github.com/vitejs/vite/pull/20767
function patchViteClientPlugin(): Plugin {
  const viteClientPath = normalizePath(
    fileURLToPath(import.meta.resolve("vite/dist/client/client.mjs")),
  );

  function endIndexOf(code: string, searchValue: string) {
    const i = code.lastIndexOf(searchValue);
    return i === -1 ? i : i + searchValue.length;
  }

  return {
    name: "fullstack:patch-vite-client",
    transform: {
      handler(code, id) {
        if (id === viteClientPath) {
          const s = new MagicString(code);
          s.prependLeft(
            code.indexOf("const sheetsMap"),
            `\
const linkSheetsMap = new Map();
document
  .querySelectorAll('link[rel="stylesheet"][data-vite-dev-id]')
  .forEach((el) => {
    linkSheetsMap.set(el.getAttribute('data-vite-dev-id'), el)
  });
`,
          );
          s.appendLeft(
            endIndexOf(code, `function updateStyle(id, content) {`),
            `if (linkSheetsMap.has(id)) { return }`,
          );
          s.appendLeft(
            endIndexOf(code, `function removeStyle(id) {`),
            `
const link = linkSheetsMap.get(id);
if (link) {
  document
    .querySelectorAll(
      'link[rel="stylesheet"][data-vite-dev-id]',
    )
    .forEach((el) => {
      if (el.getAttribute('data-vite-dev-id') === id) {
        el.remove()
      }
    })
  linkSheetsMap.delete(id)
}
`,
          );
          return s.toString();
        }
      },
    },
  };
}

// TODO: upstream?
// Vite client HMR requests scoped css link stylesheet with `lang.css=` instead of `lang.css`,
// which seems to cause response to be `content-type: text/javascript` even though response text is raw css.
// This middleware rewrites url so that Vite will add `text/css`.
function patchVueScopeCssHmr(): Plugin {
  return {
    name: "fullstack:patch-vue-scoped-css-hmr",
    configureServer(server) {
      server.middlewares.use((req, _res, next) => {
        if (
          req.headers.accept?.includes("text/css") &&
          req.url?.includes("&lang.css=")
        ) {
          req.url = req.url.replace("&lang.css=", "?lang.css");
        }
        next();
      });
    },
  };
}
