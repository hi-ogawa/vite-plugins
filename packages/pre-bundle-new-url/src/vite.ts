import type { Plugin } from "vite";
import { esbuildPluginNewUrl, esbuildPluginWorkerNewUrl } from "./esbuild";

export function vitePluginPreBundleNewUrl(): Plugin {
  return {
    name: "pre-bundle-new-url",
    config(config, _env) {
      config.cacheDir;
      return {
        optimizeDeps: {
          esbuildOptions: {
            plugins: [
              esbuildPluginWorkerNewUrl(),
              esbuildPluginNewUrl(),
            ],
          },
        },
      };
    },
    configResolved(config) {
      config.cacheDir;
    },
  };
}
