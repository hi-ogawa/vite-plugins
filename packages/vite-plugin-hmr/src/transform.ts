import type { Node, Program } from "estree";

// traverse export declaration statements based on
// https://github.com/vitejs/vite/blob/fc2bceb09fb65cc6dc843462f51506586251a703/packages/vite/src/node/ssr/ssrTransform.ts#L172

declare module "estree" {
  interface BaseNode {
    start: number;
    end: number;
  }
}

export function analyzeExports(code: string, ast: Program) {
  // extract exported top-level identifiers
  const exportIds: string[] = [];

  // unsupported export usage
  const errors: { node: Node }[] = [];

  for (const node of ast.body) {
    // named exports
    if (node.type === "ExportNamedDeclaration") {
      if (node.declaration) {
        if (
          node.declaration.type === "FunctionDeclaration" ||
          node.declaration.type === "ClassDeclaration"
        ) {
          //// export function foo() {}
          exportIds.push(node.declaration.id.name);
        } else if (node.declaration.type === "VariableDeclaration") {
          //// export const foo = 1, bar = 2
          if (node.declaration.kind === "const") {
            // rewrite from "const" to "let"
            code = edit(
              code,
              node.declaration.start,
              node.declaration.start + 5,
              "let  "
            );
          }
          for (const decl of node.declaration.declarations) {
            if (decl.id.type === "Identifier") {
              exportIds.push(decl.id.name);
            } else {
              errors.push({ node: decl });
            }
          }
        }
      } else {
        if (node.source) {
          //// export { foo, bar } from './foo'
          errors.push({ node });
        } else {
          //// export { foo, bar }
          // TODO: support by analyzing scope?
          errors.push({ node });
        }
      }
    }

    // default export
    if (node.type === "ExportDefaultDeclaration") {
      if (
        (node.declaration.type === "FunctionDeclaration" ||
          node.declaration.type === "ClassExpression") &&
        node.declaration.id
      ) {
        //// export default function foo() {}
        //// export default class A {}
        exportIds.push(node.declaration.id.name);
      } else {
        // anonymous default exports
        errors.push({ node });
      }
    }

    //// export * from './foo'
    if (node.type === "ExportAllDeclaration") {
      errors.push({ node });
    }
  }

  return { code, exportIds, errors };
}

function edit(s: string, start: number, end: number, insertee: string) {
  return s.slice(0, start) + insertee + s.slice(end);
}
