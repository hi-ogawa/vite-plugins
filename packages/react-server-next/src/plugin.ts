import path from "node:path";
import { vitePluginReactServer } from "@hiogawa/react-server/plugin";
import {
  vitePluginLogger,
  vitePluginSsrMiddleware,
} from "@hiogawa/vite-plugin-ssr-middleware";
import react from "@vitejs/plugin-react";
import type { PluginOption } from "vite";

export function vitePluginReactServerNext({
  plugins,
}: {
  plugins?: PluginOption[];
}): PluginOption {
  return [
    react(),
    vitePluginReactServer({
      plugins,
      routeDir: "app",
    }),
    // for now, ssr entry setup is not done by vitePluginReactServer
    vitePluginLogger(),
    vitePluginSsrMiddleware({
      entry: "virtual:entry-ssr",
      preview: path.resolve("./dist/server/index.js"),
    }),
    {
      name: "virtual-entry-ssr",
      resolveId(source, _importer, _options) {
        if (source === "virtual:entry-ssr") {
          return "\0" + source;
        }
        return;
      },
      load(id, _options) {
        if (id === "\0virtual:entry-ssr") {
          return `
            import { handler } from "@hiogawa/react-server/entry-server";
            import { webToNodeHandler } from "@hiogawa/utils-node";
            export default webToNodeHandler(handler);
          `;
        }
        return;
      },
    },
  ];
}
