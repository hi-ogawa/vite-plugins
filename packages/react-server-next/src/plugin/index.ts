import path from "node:path";
import { vitePluginReactServer } from "@hiogawa/react-server/plugin";
import {
  vitePluginLogger,
  vitePluginSsrMiddleware,
} from "@hiogawa/vite-plugin-ssr-middleware";
import type { Plugin, PluginOption } from "vite";

export default function next(options?: {
  plugins?: PluginOption[];
}): PluginOption {
  return [
    vitePluginReactServer({
      routeDir: "app",
      entryBrowser: "@hiogawa/react-server-next/plugin/entry-browser",
      entryServer: "@hiogawa/react-server-next/plugin/entry-server",
      buildScanMode: "server",
      plugins: [nextEsbuildJsx, ...(options?.plugins ?? [])],
    }),
    vitePluginLogger(),
    vitePluginSsrMiddleware({
      entry: "@hiogawa/react-server-next/plugin/entry-ssr",
      preview: path.resolve("./dist/server/index.js"),
    }),
  ];
}

// overrdied next.js's default `jsx: preserve`
const nextEsbuildJsx: Plugin = {
  name: "next-esbuild-jsx",
  config: () => ({
    esbuild: {
      jsx: "automatic",
    },
    optimizeDeps: { esbuildOptions: { jsx: "automatic" } },
  }),
};
