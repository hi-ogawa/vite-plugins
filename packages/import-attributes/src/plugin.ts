import * as esModuleLexer from "es-module-lexer";
import MagicString from "magic-string";
import { type Plugin, createFilter } from "vite";

export default function vitePluginImportAttributes(pluginOptions?: {
  include: string | RegExp | (string | RegExp)[];
  exclude?: string | RegExp | (string | RegExp)[];
}): Plugin[] {
  const filter = createFilter(
    pluginOptions?.include,
    pluginOptions?.exclude ?? /\/node_modules\//,
  );
  return [
    {
      name: "import-attributes",
      async config() {
        await esModuleLexer.init;
      },
      transform(code, id) {
        if (!code.includes("with")) return;
        if (!filter(id)) return;

        const transformed = transformImportAttributes(code);
        if (transformed) {
          return {
            code: transformed.toString(),
            map: transformed.generateMap({ hires: "boundary" }),
          };
        }
      },
    },
  ];
}

const KEY = "__attributes";

export function transformImportAttributes(
  code: string,
): MagicString | undefined {
  const parsed = esModuleLexer.parse(code);
  let output: MagicString | undefined;
  for (const importSpecifier of parsed[0]) {
    const { e: end, se: expEnd, a: attributeIndex } = importSpecifier;
    if (attributeIndex > -1) {
      const attributesCode = code.slice(attributeIndex, expEnd);
      const attributes = evalValue(attributesCode);
      output ??= new MagicString(code);
      output.appendLeft(
        end,
        "?" +
          new URLSearchParams({
            [KEY]: JSON.stringify(attributes),
          }),
      );
      output.remove(end + 1, expEnd);
    }
  }
  return output;
}

export function getImportAttributesFromId(id: string): object | undefined {
  const { query } = parseIdQuery(id);
  const attributes = query[KEY];
  return attributes as any;
}

function parseIdQuery(id: string): {
  filename: string | undefined;
  query: Record<string, string>;
} {
  if (!id.includes("?")) return { filename: id, query: {} };
  const [filename, rawQuery] = id.split(`?`, 2);
  const query = Object.fromEntries(new URLSearchParams(rawQuery));
  return { filename, query };
}

// https://github.com/vitejs/vite/blob/ea9aed7ebcb7f4be542bd2a384cbcb5a1e7b31bd/packages/vite/src/node/utils.ts#L1469-L1475
function evalValue<T = any>(rawValue: string): T {
  const fn = new Function(`
    var console, exports, global, module, process, require
    return (\n${rawValue}\n)
  `);
  return fn();
}
