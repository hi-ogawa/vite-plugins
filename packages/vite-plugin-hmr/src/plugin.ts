import {
  type FilterPattern,
  type Plugin,
  createFilter,
  parseAstAsync,
} from "vite";
import { name as packageName } from "../package.json";
import { analyzeExports } from "./transform";

export function vitePluginHmr(pluginOpts: {
  include?: FilterPattern;
  exclude?: FilterPattern;
}): Plugin {
  const filter = createFilter(
    pluginOpts.include,
    pluginOpts.exclude ?? ["**/node_modules/**"]
  );

  return {
    name: packageName,
    apply: "serve",
    transform(code, id, options) {
      if (options?.ssr && filter(id)) {
        return hmrTransform(code);
      }
      return;
    },
  };
}

export async function hmrTransform(code: string): Promise<string | undefined> {
  const ast = await parseAstAsync(code);
  const result = analyzeExports(code, ast as any);

  // always invalidate for hmr unsupported
  if (result.errors.length > 0) {
    const node = result.errors[0]!.node;
    const message = "unsupported usage: " + code.slice(node.start, node.end);
    return code + "\n" + generateFooterUnsupported(message);
  }

  return result.code + "\n" + generateFooter(result.exportIds);
}

function generateFooter(names: string[]) {
  const parts = names.map(
    (name) => `
  $$registry.exports["${name}"] = {
    value: ${name},
    update: ($$next) => {
      ${name} = $$next;
    }
  };
`
  );

  // requires dummy "hot.accept" for vite to detect
  return `
if (import.meta.env.SSR && import.meta.hot) {
  const $$hmr = await import("@hiogawa/vite-plugin-hmr/runtime");
  const $$registry = $$hmr.createRegistry();

${parts.join("\n")}

  $$hmr.setupHot(import.meta.hot, $$registry);
  import.meta.hot.accept;
}
`;
}

function generateFooterUnsupported(message: string) {
  return `
if (import.meta.env.SSR && import.meta.hot) {
  import.meta.hot.accept(() => {
    import.meta.hot.invalidate(${JSON.stringify(message)})
  });
}
`;
}
