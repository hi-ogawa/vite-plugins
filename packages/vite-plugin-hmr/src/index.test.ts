import MagicString from "magic-string";
import { parseAstAsync } from "vite";
import { describe, expect, it } from "vitest";
import { hmrTransform } from "./plugin";
import { analyzeExports } from "./transform";

describe(hmrTransform, () => {
  const examples = [
    [
      "basic",
      `
export function Test1() {}

export let Test2 = () => {}

export /*some*/ const /*comment*/ Test3 = () => {}

export default function Test4() {}
`,
    ],
    [
      "unsupported separate named export",
      `
const SomeNamed = () => {};
export { SomeNamed };
`,
    ],
    [
      "unsupported separate default export",
      `
const SomeDeafult = () => {};
export default SomeDefault;
`,
    ],
  ] as const;

  examples.forEach(([title, code], i) => {
    describe(`${i} - ${title}`, () => {
      it(analyzeExports, async () => {
        const ast = await parseAstAsync(code);
        const { exportIds, errors } = analyzeExports(
          new MagicString(code),
          ast as any
        );
        expect({
          exportIds,
          errors: errors.map((e) => code.slice(e.node.start, e.node.end)),
        }).toMatchSnapshot();
      });

      it(hmrTransform, async () => {
        expect(await hmrTransform(code)).toMatchSnapshot();
      });
    });
  });
});
