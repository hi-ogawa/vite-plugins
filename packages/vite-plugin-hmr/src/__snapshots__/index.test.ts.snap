// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hmrTransform > 0 - basic > analyzeExports 1`] = `
{
  "errors": [],
  "exportIds": [
    "Test1",
    "Test2",
    "Test3",
    "Test4",
  ],
}
`;

exports[`hmrTransform > 0 - basic > hmrTransform 1`] = `
{
  "code": "
export function Test1() {}

export let Test2 = () => {}

export /*some*/ let /*comment*/ Test3 = () => {}

export default function Test4() {}


if (import.meta.env.SSR && import.meta.hot) {
  const $$hmr = await import("@hiogawa/vite-plugin-hmr/runtime");
  const $$registry = $$hmr.createRegistry();


  $$registry.exports["Test1"] = {
    value: Test1,
    update: ($$next) => {
      Test1 = $$next;
    }
  };


  $$registry.exports["Test2"] = {
    value: Test2,
    update: ($$next) => {
      Test2 = $$next;
    }
  };


  $$registry.exports["Test3"] = {
    value: Test3,
    update: ($$next) => {
      Test3 = $$next;
    }
  };


  $$registry.exports["Test4"] = {
    value: Test4,
    update: ($$next) => {
      Test4 = $$next;
    }
  };


  $$hmr.setupHot(import.meta.hot, $$registry);
  import.meta.hot.accept;
}
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA,MAAM,CAAC,QAAQ,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,CAAC;AAC1B;AACA,MAAM,CAAC,GAAG,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAC3B;AACA,MAAM,CAAC,CAAC,CAAC,IAAI,CAAC,CAAC,IAAM,CAAC,CAAC,CAAC,OAAO,CAAC,CAAC,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAClD;AACA,MAAM,CAAC,OAAO,CAAC,QAAQ,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,CAAC;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`hmrTransform > 1 - unsupported separate named export > analyzeExports 1`] = `
{
  "errors": [
    "export { SomeNamed };",
  ],
  "exportIds": [],
}
`;

exports[`hmrTransform > 1 - unsupported separate named export > hmrTransform 1`] = `
{
  "code": "
const SomeNamed = () => {};
export { SomeNamed };


if (import.meta.env.SSR && import.meta.hot) {
  import.meta.hot.accept(() => {
    import.meta.hot.invalidate("unsupported usage: export { SomeNamed };")
  });
}
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA,KAAK,CAAC,SAAS,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAC3B,MAAM,CAAC,CAAC,CAAC,SAAS,CAAC,CAAC,CAAC;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`hmrTransform > 2 - unsupported separate default export > analyzeExports 1`] = `
{
  "errors": [
    "export default SomeDefault;",
  ],
  "exportIds": [],
}
`;

exports[`hmrTransform > 2 - unsupported separate default export > hmrTransform 1`] = `
{
  "code": "
const SomeDeafult = () => {};
export default SomeDefault;


if (import.meta.env.SSR && import.meta.hot) {
  import.meta.hot.accept(() => {
    import.meta.hot.invalidate("unsupported usage: export default SomeDefault;")
  });
}
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA,KAAK,CAAC,WAAW,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAC7B,MAAM,CAAC,OAAO,CAAC,WAAW,CAAC;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
